text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1412 @@
+they aren't `Sized`.
+
+Introducing a `const Sized` trait will enable `Copy` to be implemented for
+those types whose size is a runtime constant to function correctly without
+special cases in the type system. See
+[rfcs#3268: Scalable Vectors][rfc_scalable_vectors].
+
+## Unsized types and extern types
+[unsized-types-and-extern-types]: #unsized-types-and-extern-types
+
+[Extern types][rfc_extern_types] has long been blocked on these types being
+neither `Sized` nor `?Sized` ([relevant issue][issue_extern_types_align_size]).
+
+RFC #1861 defined that `std::mem::size_of_val` and `std::mem::align_of_val`
+should not be defined for extern types but not how this should be achieved, and
+suggested an initial implementation could panic when called with extern types,
+but this is always wrong, and not desirable behavior in keeping with Rust's
+values. `size_of_val` returns 0 and `align_of_val` returns 1 for extern types in
+the current implementation. Ideally `size_of_val` and `align_of_val` would error
+if called with an extern type, but this cannot be expressed in the bounds of
+`size_of_val` and `align_of_val` and this remains a blocker for extern types.
+
+Furthermore, unsized types cannot be members of structs as their alignment is
+unknown and this is necessary to calculate field offsets. Extern types also
+cannot be used in `Box` as `Box` requires size and alignment for both allocation
+and deallocation.
+
+Introducing a hierarchy of `Sized` traits will enable the backwards-compatible
+introduction of a trait which only extern types do not implement and will
+therefore enable the bounds of `size_of_val` and `align_of_val` to disallow
+instantiations with extern types.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+Most types in Rust have a size known at compilation time, such as `u32` or
+`String`. However, some types in Rust do not have known sizes.
+
+For example, slices have an unknown length while compiling and are
+known as *dynamically-sized types*, their size must be computed at runtime.
+There are also types with a runtime constant size (but unknown at compile time),
+and *unsized types* with no size whatsoever.
+
+Various parts of Rust depend on knowledge of the size of a type to work, for
+example:
+
+- [`std::mem::size_of_val`][api_size_of_val] computes the size of a value,
+  and thus cannot accept extern types which have no size, and this should
+  be prevented by the type system.
+- Rust allows dynamically-sized types to be used as the final field in a struct,
+  but the alignment of the type must be known, which is not the case for extern
+  types.
+- Allocation and deallocation of an object with `Box` requires knowledge of
+  its size and alignment, which extern types do not have.
+- For a value type to be allocated on the stack, it needs to have constant
+  known size, which dynamically-sized and unsized types do not have (but 
+  sized and "runtime-sized" types do).
```

Comment:
We might want to clarify that technically speaking dynamic stack allocation *exists* (cc `alloca`, C's VLAs) just not in Rust (barring incomplete features like `unsized_locals`, `unsized_fn_params`).

Reply:
Added some wording about this.